### ISSUE TITLE
sbt fails to locate an sbt plugin from private Nexus in Maven mode, when Nexus enforces consistency

### ISSUE THREAD

**Issue by cchepelov** [sbt/sbt#6384]
## steps

1. publish a plugin to a Sonatype Nexus instance. 

  * Due to corporate policy which enforces strict Maven consistency in URI paths, the plugin build includes traditional steps such as 
  ```
  ThisBuild / pomIncludeRepository := { _ => false },
  ThisBuild / publishMavenStyle := true 
  ```   
  
but it also includes various tweaks to `moduleSettings`, `artifactName` and `pomPostProcess`, in order to publish to 
   https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin_2.12_1.0-1.2.3.pom
(instead of ..../my-plugin-1.2.3.pom), while keeping `<artifactId>my-plugin</artifactId>` (without suffixes) in the pom itself.

Doing these tweaks enables the plugin to be delivered and pass the strict consistency checks mandated by $EMPLOYER (non-negotiable at this moment).

2. attempt to use the plugin
   
```
addSbtPlugin("org.path.to" %% "my-plugin" % "1.2.3")
``` 
--> not found: https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin-1.2.3.pom 

(which is fair as we explicitly attempted to not publish there, but a bit frustrating as perhaps sbt might also look at the "suffixed" pom name as well?)

or 

```
addSbtPlugin("org.path.to" % "my-plugin" % "1.2.3")
// notice the single % 
``` 
--> not found: https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin-1.2.3.pom 

Why did sbt append the _2.12_1.0 suffix while we told it not to ? 

```
addSbtPlugin("org.path.to" % "my-plugin_2.12_1.0" % "1.2.3")
// notice the single % 
``` 
--> not found: https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0_2.12_1.0/1.2.3/my-plugin_2.12_1.0-1.2.3.pom 
aha, sbt added the suffix even though we did it manually :(

```
addSbtPlugin("org.path.to" %% "my-plugin" % "1.2.3" from "https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0_2.12_1.0/1.2.3/my-plugin_2.12_1.0-1.2.3.jar")
``` 
--> not found: https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin-1.2.3.pom 
the "from" clause from ye olde Ivy times got ignored 

Additional attempts including declaring non-standard patterns in the `repositories` file failed as well (as soon as `mavenCompatible` is declared, the pattern is ignored. If the pattern is declared, sbt looks for an Ivy file which we can't publish due to Sonatype Nexus and/or corporate policy)

## problem

the problem is that sbt looks for plugins into a path that Sonatype Nexus can consider as invalid, and there appears to be no way to override that.

## expectation

There is a way to publish an sbt plugin to a "strict mode" Sonatype Plugin and consume it in Maven mode

A. 
when declaring `addSbtPlugin("org.path.to" %% "my-plugin" % "1.2.3")`, sbt looks for https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin-1.2.3.pom  (traditional) and https://nexus.private/repo/org/path/to/my-plugin_2.12_1.0/1.2.3/my-plugin_2.12_1.0-1.2.3.pom (forced consistency)

B. when declaring `addSbtPlugin("org.path.to" % "my-plugin_2.12_1.0" % "1.2.3")`, sbt obeys and doesn't slap any additional suffixes

C. sbt obeys the `from url` syntax again especially in the addSbtPlugin clause.

D. even in mavenMode, sbt obeys the pattern declared in "repositories" should the user know better.

**Comment by cchepelov** [comment]
Found a workaround !!!!!

```
libraryDependencies += ("org.path.to" % "my-plugin_2.12_1.0" % "1.2.3") 
  // in addition to the funky workaround at the publish side
```

it was kind of obvious by looking at the surprisingly simple code under addSbtPlugin()'s implementation. Shoulda done this earlier.

**Comment by eed3si9n** [comment]
@cchepelov Did you wipe out the extra somehow when you published it? This might be a good workaround to document since we publish sbt major versions so infrequently.

**Comment by cchepelov** [comment]
Yes, I need to go back to my plug-in, clean up and minimize the mess, and 
will document that.

**Comment by loicdescotte** [comment]
I have the same issue, the workaround does not work  for me (with sbt-pack plugin). 
@cchepelov did you keep the plugins.sbt line too, in addition to the libraryDependencies  ?

**Comment by cchepelov** [comment]
To document, the solution I found:

In the plugin:
```scala
lazy val root = (projet in file("."))
 .settings(
    /// blah
    // traditional Maven:
    ThisBuild / pomIncludeRepository := { _ => false }, 
    ThisBuild / publishMavenStyle := true,
    // the less traditional:
    ThisBuild / publishTo := {
       val nexus = "http://nexus.corporate/repository/team/"
       val (name, baseUrl) = if (isSnapshot.value) 
          ("PROJECT Snapshots", nexus + "project-snapshot")
       else 
          ("PROJECT Releases", nexus + "project-release")
           
       val pattern = "[organisation]/[module]_[scalaVersion]_[sbtVersion]/[revision]/[artifact]_[scalaVersion]_[sbtVersion]-[revision](-[classifier]).[ext]"
       
       val targetRepo = URLRepository(name, Patterns()
            .withArtifactPattern(Vector(baseUrl + "/" + pattern))
            .withIsMavenCompatible(true)

       Some(targetRepo)          
    }
    )
```

and in the consumer side, in `projects/plugins.sbt`; rather than "addSbtPlugin":

```scala
libraryDependencies += "corp.team" % "sbt-project-plugin_2.12_1.0" % "1.0.0"    
   // notice the use of libraryDependencies, single % and manual suffixes  
```

I hope this helps; please report whether it works for you @loicdescotte

**Comment by loicdescotte** [comment]
@cchepelov thanks, I thought that you added the libraryDependencies workaround in build.sbt

I also needed to add this in plugins.sbt (like I did in build.sbt):

resolvers +=  "Internal Nexus Repo" at "https://nexus.company.com:8080/repository/maven-releases"

Now the URL resolution is working, thanks, but I have a plugin binary compatibility error, I don't know if it can be related to the workaround or not (I never had this issue before using a private repo, using Scala 2.13 in my code but plugins seem to use 2.12) : 

````
sbt.internal.IncompatiblePluginsException: Binary incompatibility in plugins detected.

Caused by: java.lang.NoClassDefFoundError: org/apache/commons/compress/compressors/gzip/GzipCompressorOutputStream

```

**Comment by loicdescotte** [comment]
The workaround is working for me too now, I had to add manually some transitive plugin dependencies.

plugins.sbt : 


```
resolvers +=  "Nexus Internal Repo" at "https://nexus.xxx:8080/repository/maven-releases"  

libraryDependencies += "org.apache.commons" % "commons-compress" % "1.20"
libraryDependencies += "com.typesafe.play" % "twirl-api_2.12" % "1.5.0"  

libraryDependencies += "org.xerial.sbt" % "sbt-pack_2.12_1.0" % "0.13"
```
